README.md
Server side application skeleton
===
![caution](https://user-images.githubusercontent.com/6298396/39394204-c8a60516-4ad7-11e8-8f07-5c28de190586.png)

Basic server side web application with the following functionality:
- HTTP/HTTPS servers
- routes and authentication on middleware (use [gorilla mux](https://github.com/gorilla/mux))
- session management (use [gorilla sessions](https://github.com/gorilla/sessions))
- user management (registration, change and delete user, password recovery by mail)
- offline fallback page if network fail
- modular design, can be extended very easy
- config file in JSON format
- database generated from file or model (with [Mysql Workbench](https://dev.mysql.com/downloads/workbench/))
- responsive interface for mobile access, browser back button disabled
- salted and hashed passwords in database

### Integrations

The skeleton is based on [W3CSS](https://www.w3schools.com/w3css/), a very lightweight CSS framework and is designed to integrate with [JQuery](https://jquery.com/), [Angular](https://angular.io/), [Fontawesome](https://fontawesome.com/v4.7.0/icons/) and some useful libraries like [Datetime picker](https://trentrichardson.com/examples/timepicker/) and [Google Charts](https://developers.google.com/chart/) (see `ui/head.html` file).

### Compilation modes

#### Plaint text passwords

	go build -tags="plain"

Passwords are stored in plain text in the database and sent as plain text on mail to recovery (unrecommended).

#### Hashed passwords

	go build -tags="sha1 random"

Passwords are stored hashed in the database and sent as random on mail to recovery (recommended).

#### Salted and hashed passwords

	go build -tags="saltsha1 random"

Passwords are stored salted and hashed in the database and sent as random on mail to recovery (highly recommended).

The tag `random` mean send password as random string.

### How is look like?

#### Signup screen
![signup](./static/images/signup.png)
#### Mail/reset my password
![reset](./static/images/reset.png)
#### Signin screen
![signin](./static/images/signin.png)
#### Basic home screen
![home](./static/images/home.png)
#### Local account management
![user](./static/images/user.png)
#### Network fail page
![cache](./static/images/cache.png)

### Configuration file

	{
	    "database": {
	        "ip": "example.com",
	        "port": 3306,
	        "user": "root",
	        "password": "password",
	        "name": "skel"
	    },
	    "smtp": "smtp.gmail.com",
	    "port": 587,
	    "user": "mailer@gmail.com",
	    "password": "password",
	    "name": "mailer name"
	}

Database must be created by hand before running the program. Mailer is used to recovery forgotten passwords.

### Database

Database structure is described in `database.sql` file. Adjust the database name and import the file in Mysql. Alternatively you can use the model from `model/skel_full.mwb` (with [Mysql Workbench](https://dev.mysql.com/downloads/workbench/))

### Parameters

Start the program with the following parameters:

#### `-http`

Listening address and port for HTTP server. Default `:8080`.

#### `-https`

Listening address and port for HTTPS server. Default `:8090`.

#### `-https-enabled`

Enable HTTPS server. Default `false`.

### Adding a simple application

Add menu option in `ui/navbar.html` file

	<a href="/a/dummy" class="w3-bar-item" title="Dummy page">Page</i></a>

Add route to the application in `main.go` file

	application.HandleFunc("/dummy", dummy)

Create `ui/dummy.html` file for application user interface

	[[define "dummy"]]
	<!DOCTYPE html>
	<html lang="en">
	<head>
	<title></title>
	[[template "head"]]
	</head>
	<body>
	[[template "navbar" .]]
	<div class="w3-display-container" style="height:300px;">
	  <div class="w3-display-middle w3-center">
	  <form method="post" action="">
	  <button type="submit" name="submit">Press this button!</button>
	  <p>[[.Message]]
	  </form>
	</div>
	</div>
	</body>
	</html>
	[[end]]

Create `ui_dummy.go` file who handle the application

	package main
	
	import (
		"net/http"
	)
	
	func dummy(w http.ResponseWriter, r *http.Request) {
		session, _ := store.Get(r, SESSID)
		user := session.Values["user"].(User)
		params := struct {
			Name    string // this is mandatory
			Message string // TODO add fields as you need
		}{Name: user.Name}
		switch r.Method {
		case "GET":
			templ.ExecuteTemplate(w, "dummy", params)
		case "POST":
			params.Message = "You pressed dummy button!"
			templ.ExecuteTemplate(w, "dummy", params)
		}
	}

Take care to the template parameters. Add them as you need. `Name` parameter must appear in all applications because is the user name from the top right navbar. 

Look carefully over **TODO** and **NOTE** comments because you may need to adjust as you need.

### Make your application look better with Bootstrap

For keeping things simple, the skeleton comes with a very basic set of widgets based on W3CSS. You can use [Bootstrap](https://getbootstrap.com/) following the next steps:

- remove inputs `height` from `style.html` because is useful just for W3CSS framework

		select,input,button {
		  -height:30px; /* all inputs must have the same height */

- use fixed length in forms containers, eg:

		<div class="w3-container" style="width:400px">

- add bootstrap classes on inputs (or other elements), eg:

		<input class="form-control" type="text">
		<button class="btn btn-primary" type="submit">

### Known issues

The skeleton is designed to work with HTML5 so make sure your browser support this.

CHANGES
1.0.0-release-build230419
first release

1.1.0-release-build300419
- salted and hashed passwords in database
- send random recovery passwords on mail

1.1.1-release-build030519
- reset the password to random only if the mail was successfully sent

1.1.2-release-build160619
- sticky navbar in ui

1.2.0-release-build290619
- add bootstrap support

1.2.1-release-build160719
- add vendor folder
- navbar add w3-margin-bottom
- head add shortcut icon

main.go
package main

import (
	"crypto/tls"
	"encoding/gob"
	"flag"
	"fmt"
	"log"
	"net/http"
	"os"
	"path/filepath"
	"runtime"
	"text/template"

	"github.com/geosoft1/json"
	"github.com/gorilla/mux"
	"github.com/gorilla/sessions"
)

// TODO application version
// https://semver.org/#semantic-versioning-200
const SW_VERSION = "1.2.0-release-build290619"

// NOTE session cookie name
const SESSID = "SESSID"

// NOTE cache manifest file
const MANIFEST = `CACHE MANIFEST
FALLBACK:
/ /static/offline.html
`

type User struct {
	Id       int    `json:"id"`
	Name     string `json:"name"`
	Email    string `json:"email"`
	Password string `json:"password"`
	isActive bool   `json:"is_active"`
}

type Database struct {
	Ip       string `json:"ip"`
	Port     int    `json:"port"`
	User     string `json:"user"`
	Password string `json:"password"`
	Name     string `json:"name"`
}

type SMTP struct {
	Server   string `json:"smtp"`
	Port     int    `json:"port"`
	User     string `json:"user"`
	Password string `json:"password"`
	Name     string `json:"name"`
}

// main configuration structure
type Config struct {
	Database `json:"database"`
	SMTP
}

var httpAddress = flag.String("http", ":8080", "http address")
var httpsAddress = flag.String("https", ":8090", "https address")
var httpsEnabled = flag.Bool("https-enabled", false, "enable https server")

var config = &Config{}
var templ = template.New("templ").Delims("[[", "]]") // integrate with angular
var router = mux.NewRouter()                         // main router

// NOTE this is the salt for secured passwords
const salt = "super-secret-key"

// NOTE random recovery password length
const token_len = 8

var (
	// key must be 16, 24 or 32 bytes long (AES-128, AES-192 or AES-256)
	key   = []byte(salt)
	store = sessions.NewCookieStore(key)
)

func main() {
	// NOTE this will avoid processor overload in some circumstances
	runtime.GOMAXPROCS(1)

	flag.Usage = func() {
		fmt.Printf("usage: %s [options]\n", filepath.Base(os.Args[0]))
		flag.PrintDefaults()
	}
	flag.Parse()
	log.SetFlags(log.LstdFlags | log.Lshortfile)

	folder, err := filepath.Abs(filepath.Dir(os.Args[0]))
	if err != nil {
		log.Fatalln(err)
	}

	file, err := os.Open(filepath.Join(folder, "config.json"))
	if err != nil {
		log.Fatalln(err)
	}
	json.Decode(file, &config)
	// NOTE [DEBUG] print configuration
	// log.Println(config)

	if _, err := templ.ParseGlob(filepath.Join(folder, "ui", "*.html")); err != nil {
		log.Fatalln(err)
	}

	sqlConnect()

	//https://github.com/gorilla/sessions/issues/58#issuecomment-154217736
	gob.Register(User{})

	router.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		templ.ExecuteTemplate(w, "index", nil)
	})
	router.HandleFunc("/signin", signin)
	router.HandleFunc("/signup", signup)
	router.HandleFunc("/reset", reset) // reset password if you forgot
	router.HandleFunc("/signout", signout)
	router.HandleFunc("/cache.manifest", func(w http.ResponseWriter, r *http.Request) {
		// offline page is shown if network is down
		w.Header().Set("Content-Type", "text/cache-manifest")
		w.Write([]byte(MANIFEST))
	})
	application := router.PathPrefix("/a").Subrouter()
	application.Use(func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			// resolve cross-site access issues
			w.Header().Set("Access-Control-Allow-Origin", "*")
			session, _ := store.Get(r, SESSID)
			if session.IsNew || session.Values["authenticated"] == false {
				http.Error(w, "Session expired", http.StatusOK)
				return
			}
			next.ServeHTTP(w, r)
		})
	})
	application.HandleFunc("/home", home)
	application.HandleFunc("/user", user)

	// TODO application handlers come here

	// NOTE [DEBUG] print registered routes
	router.Walk(func(route *mux.Route, router *mux.Router, ancestors []*mux.Route) error {
		// if name, err := route.GetPathTemplate(); err == nil {
		// 	log.Println(name)
		// }
		return nil
	})

	// file server
	router.PathPrefix("/static/").Handler(http.StripPrefix("/static/", http.FileServer(http.Dir(filepath.Join(folder, "static")))))
	// --- error handlers (NotFoundHandler, MethodNotAllowedHandler)
	router.NotFoundHandler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusNotFound)
		//http.Error(w, http.StatusText(http.StatusNotFound), http.StatusNotFound)
	})
	router.MethodNotAllowedHandler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusMethodNotAllowed)
	})

	if *httpsEnabled {
		go func() {
			// allow you to use self signed certificates
			http.DefaultTransport.(*http.Transport).TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
			// NOTE geneate self signed certificates for tests
			// openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout server.key -out server.crt
			log.Printf("start https server on %s", *httpsAddress)
			if err := http.ListenAndServeTLS(*httpsAddress, filepath.Join(folder, "server.crt"), filepath.Join(folder, "server.key"), router); err != nil {
				log.Fatalln(err)
			}
		}()
	}

	log.Printf("start http server on %s", *httpAddress)
	if err := http.ListenAndServe(*httpAddress, router); err != nil {
		log.Fatalln(err)
	}
}
